builder/src/lib.rs

```diff
@@ -1,10 +1,247 @@
+            }
+        } else {
+            quote! {
+                pub fn #ident(&mut self, #ident: #ty) -> &mut Self {
+                    self.#ident = std::option::Option::Some(#ident);
+                    self
+                }
+            }
+        }
+    });
+    quote! { #(#setters)* }
+}
+
+fn expand_build(name: &Ident, fields: &syn::FieldsNamed, optional_fields: &HashSet<String>, builders: &HashMap<String, String>) -> TokenStream2 {
+    let validation = fields.named.iter().filter(|f| {
+        let identstr = f.ident.as_ref().map(|x| format!("{}", x)).unwrap_or("".to_owned());
+        !optional_fields.contains(&identstr) && !builders.contains_key(&identstr)
+    }).map(|f| {
+        let ident = &f.ident;
+        let identstr = f.ident.as_ref().map(|x| format!("{}", x)).unwrap_or("".to_owned());
+        quote!{
+            if self.#ident.is_none() {
+                return Err(<std::boxed::Box<dyn std::error::Error>>::from(format!("{} is missing", #identstr)));
+            }
+        }
+    });
+    let f = fields.named.iter().map(|f| {
+        let ident = &f.ident;
+        let identstr = f.ident.as_ref().map(|x| format!("{}", x)).unwrap_or("".to_owned());
+        if optional_fields.contains(&identstr) {
+            quote! {
+                #ident: self.#ident.take(),
+            }
+        } else if builders.contains_key(&identstr) {
+            quote! {
+                #ident: self.#ident.drain(..).collect(),
+            }
+        } else {
+            quote! {
+                #ident: self.#ident.take().unwrap(),
+            }
+        }
+    });
+    quote! {
+        #(#validation)*
+        Ok(#name {
+            #(#f)*
+        })
+    }
+}
+
+fn get_optional_fields(fields: &syn::FieldsNamed) -> HashSet<String> {
+    fields.named.iter().filter(|f| {
+        match f.ty {
+            syn::Type::Path(ref path) => {
+                &path.path.segments[0].ident == "Option"
+            },
+            _ => false,
+        }
+    }).map(|f| {
+        f.ident.as_ref().map(|x| format!("{}", x)).unwrap()
+    }).collect()
+}
+
+fn get_fields_builders(fields: &syn::FieldsNamed) -> Option<HashMap<String, String>> {
+    let builderfields = fields.named.iter().filter(|f| {
+        f.attrs.len() > 0 && f.attrs.iter().any(|attr| &attr.path.segments[0].ident.to_string() == "builder")
+    }).collect::<Vec<_>>();
+    let hm = builderfields.iter().filter_map(|f| {
+        let attr = f.attrs.iter().find(|attr| &attr.path.segments[0].ident.to_string() == "builder").unwrap();
+        let meta = attr.parse_meta().unwrap();
+        let name = match meta {
+            syn::Meta::List(l) => {
+                l.nested.into_iter().find_map(|m| {
+                    match m {
+                        syn::NestedMeta::Meta(ref m) => {
+                            match m {
+                                syn::Meta::NameValue(ref kv) => {
+                                    if &kv.ident.to_string() == "each" {
```

Comment:
We provide the right PartialEq impl on Ident such that this can be written more concisely as:

```rust
if kv.ident == "each" {
```